File: Backend/Financeiro.ConciliacaoWebAPI/JsonWebToken/Code/GestaoCarga/GestaoCargaDAL.cs
using JsonWebToken.Model;
using System;
using System.Collections.Generic;
using Oracle.ManagedDataAccess.Client;
using JsonWebToken.Util;

namespace JsonWebToken.Code.GestaoCarga
{
  public class GestaoCargaDAL
  {

    public OracleConnection oracleConnection;
    public GestaoCargaDAL()
    {
      oracleConnection = new OracleConnection(Utils.ORACLE_CONNECCAO);
      oracleConnection.Open();
    }


    public GestaoCargaModel selectGestaoCargaPorId(int id)
    {

      GestaoCargaSQL sql = new GestaoCargaSQL();


      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      command.CommandText = sql.selectGestaoCargaPorId();

      command.Parameters.Add("codigo", id);
      command.CommandType = System.Data.CommandType.Text;
      OracleDataReader reader = command.ExecuteReader();
      GestaoCargaModel gestaoCarga = null;

      try
      {
        if (reader.Read())
        {
          gestaoCarga = new GestaoCargaModel { codGestaoCarga = reader.GetInt32(0), data = reader.GetDateTime(1), dataMovimentacao = reader.GetDateTime(2), tipo = reader.GetString(3), versao = reader.GetInt32(4), versaoOficial = reader.GetString(5), valor = reader.GetFloat(6), versaoOficialLabel = reader.GetString(5) == "1" ? "Sim" : "Não" };
        }


        return gestaoCarga;
      }
      catch (Exception ex)
      {
        throw ex;
      }
      finally
      {

        if (oracleConnection != null)
          oracleConnection.Close();
      }
    }
    public void deleteGestaoCarga(int id)
    {

      OracleCommand command = new OracleCommand();
      OracleTransaction transaction = oracleConnection.BeginTransaction();
      // Assign transaction object for a pending local transaction
      command.Connection = oracleConnection;
      command.Transaction = transaction;
      try
      {
        GestaoCargaSQL sql = new GestaoCargaSQL();
        command.CommandText = sql.deleteGestaoCargaCtb();
        command.Parameters.Add("codigo", id);

        LogUtil.Execute(new Dictionary<string, object>(), "MOVCTBCLOSYSCOB", command); // antes = delete

        command.ExecuteNonQuery();
        command.CommandText = sql.deleteGestaoCargaTit();
        command.ExecuteNonQuery();
        command.CommandText = sql.deleteGestaoCargaAgn();
        command.ExecuteNonQuery();
        command.CommandText = sql.deleteGestaoCarga();
        command.ExecuteNonQuery();
        transaction.Commit();
      }
      catch (Exception ex)
      {
        transaction.Rollback();

        throw ex;
      }
    }

    public List<GestaoCargaModel> selectGestaoCarga(DateTime dataRef, String tipo)
    {
      GestaoCargaSQL sql = new GestaoCargaSQL();


      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      command.CommandText = sql.selectGestaoCarga();

      command.Parameters.Add("dataref", dataRef.ToString("yyyy-MM-dd"));
      command.Parameters.Add("tipo", tipo.Trim().ToUpper());
      command.CommandType = System.Data.CommandType.Text;
      OracleDataReader reader = command.ExecuteReader();
      List<GestaoCargaModel> gestaoCarga = new List<GestaoCargaModel>();

      try
      {

        while (reader.Read())
        {
          gestaoCarga.Add(new GestaoCargaModel { codGestaoCarga = reader.GetInt32(0), data = reader.GetDateTime(1), dataMovimentacao = reader.GetDateTime(2), tipo = reader.GetString(3), versao = reader.GetInt32(4), versaoOficial = reader.GetString(5), valor = reader.GetFloat(6), versaoOficialLabel = reader.GetString(5) == "1" ? "Sim" : "Não" });
        }


        return gestaoCarga;
      }
      catch (Exception ex)
      {
        throw ex;
      }
      finally
      {

        if (oracleConnection != null)
          oracleConnection.Close();
      }
    }

    private void updateGestaoCarga(GestaoCargaModel gestaoCargaModel, OracleCommand command)

    {
      GestaoCargaSQL sql = new GestaoCargaSQL();

      command.CommandText = sql.updateGestaoCargaVersaoOficialAtivar();
      command.Parameters.Clear();
      command.Parameters.Add("codigoUpdate", gestaoCargaModel.codGestaoCarga);

      command.ExecuteNonQuery();
      command.CommandText = sql.updateGestaoCargaVersaoOficialInativar();
      command.Parameters.Add("datarefUpdate", gestaoCargaModel.data.ToString("yyyy-MM-dd"));
      command.Parameters.Add("tipoUpdate", gestaoCargaModel.tipo);

      var dic = new Dictionary<string, object>();
      dic.Add("IDTVRSRGTATV", 88);

      LogUtil.Execute(dic, "RLCRGTMOVCLO", command); // antes = update

      command.ExecuteNonQuery();

    }

    public void insertGestaoCarga(DateTime dataRef)
    {

      inserirLoteTitulo();
      inserirLoteContabil();
      inserirLoteAging();

    }
    public int proximoCodigoGestaoCarga(OracleCommand command)
    {
      GestaoCargaSQL sql = new GestaoCargaSQL();
      command.CommandText = sql.proximoCodigoGestaoCarga();
      command.CommandType = System.Data.CommandType.Text;
      OracleDataReader reader = command.ExecuteReader();
      int codGestaoCarga = 0;
      if (reader.Read())
      {
        codGestaoCarga = reader.GetInt32(0);
      }
      return codGestaoCarga;
    }
    private void inserirLoteAging()
    {
      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      OracleTransaction transaction;

      // Start a local transaction
      transaction = oracleConnection.BeginTransaction();
      // Assign transaction object for a pending local transaction
      command.Transaction = transaction;

      try
      {

        GestaoCargaSQL sql = new GestaoCargaSQL();

        GestaoCargaModel gestaoCargaModel = gerarGestaoCargaModel("AGN", command);


        command.CommandText = sql.insertLoteAging();
        command.CommandType = System.Data.CommandType.Text;
        command.Parameters.Clear();
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga);
        command.Parameters.Add("datamov", gestaoCargaModel.data);
        command.ExecuteNonQuery();
        transaction.Commit();
      }
      catch (Exception ex)
      {
        transaction.Rollback();

        throw ex;
      }
    }
    public GestaoCargaModel gerarGestaoCargaModel(string tipo, OracleCommand command)
    {

      GestaoCargaModel gestaoCargaModel = new GestaoCargaModel();
      gestaoCargaModel.codGestaoCarga = proximoCodigoGestaoCarga(command);
      gestaoCargaModel.tipo = tipo;
      gestaoCargaModel.data = DateTime.Now;
      gestaoCargaModel.dataMovimentacao = DateTime.Now;

      gestaoCargaModel.versao = selectMaxVersao(gestaoCargaModel.data, gestaoCargaModel.tipo, command);
      GestaoCargaSQL sql = new GestaoCargaSQL();
      command.CommandText = sql.insertGestaoCarga();
      command.Parameters.Clear();
      command.Parameters.Add("codigo", gestaoCargaModel.codGestaoCarga);
      command.Parameters.Add("datamov", gestaoCargaModel.dataMovimentacao);
      command.Parameters.Add("tipo", gestaoCargaModel.tipo);
      command.Parameters.Add("versao", gestaoCargaModel.versao);
      command.ExecuteNonQuery();
      updateGestaoCarga(gestaoCargaModel, command);
      return gestaoCargaModel;
    }

    private int selectMaxVersao(DateTime data, string tipo, OracleCommand command)
    {

      GestaoCargaSQL sql = new GestaoCargaSQL();
      command.CommandText = sql.selectMaxVersion();
      command.CommandType = System.Data.CommandType.Text;
      command.Parameters.Add("dataRefVer", data.ToString("yyyy-MM-dd"));
      command.Parameters.Add("tipoVer", tipo);
      OracleDataReader reader = command.ExecuteReader();
      int versao = 0;
      if (reader.Read())
      {
        versao = reader.GetInt32(0);
      }
      return versao;
    }

    private void inserirLoteContabil()
    {

      OracleTransaction transaction;

      // Start a local transaction
      transaction = oracleConnection.BeginTransaction();
      // Assign transaction object for a pending local transaction
      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      command.Transaction = transaction;

      try
      {
        GestaoCargaSQL sql = new GestaoCargaSQL();
        GestaoCargaModel gestaoCargaModel = gerarGestaoCargaModel("CTB", command);

        command.CommandText = sql.insertLoteContabil();
        command.CommandType = System.Data.CommandType.Text;
        command.Parameters.Clear();
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga);
        command.Parameters.Add("datamov", gestaoCargaModel.data);
        command.ExecuteNonQuery();
        transaction.Commit();
      }
      catch (Exception ex)
      {
        transaction.Rollback();

        throw ex;
      }
    }

    private void inserirLoteTitulo()
    {


      OracleTransaction transaction;

      // Start a local transaction
      transaction = oracleConnection.BeginTransaction();
      // Assign transaction object for a pending local transaction
      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      command.Transaction = transaction;

      try
      {
        GestaoCargaSQL sql = new GestaoCargaSQL();

        GestaoCargaModel gestaoCargaModel = gerarGestaoCargaModel("TIT", command);

        command.CommandText = sql.insertLoteTitulo();
        command.CommandType = System.Data.CommandType.Text;
        command.Parameters.Clear();
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga);
        command.Parameters.Add("datamov", gestaoCargaModel.data);
        command.ExecuteNonQuery();

        LogUtil.Execute(new Dictionary<string, object>(), "MOVTITCLOSYSCOB", command, "where NUMSEQCRGTAB = :Codigo"); // depois = insert

        transaction.Commit();
      }
      catch (Exception ex)
      {
        transaction.Rollback();

        throw ex;
      }
    }
    public void atualizarVersao(GestaoCargaModel gestaoCargaModel)
    {
      OracleCommand command = new OracleCommand();
      command.Connection = oracleConnection;
      OracleTransaction transaction;

      // Start a local transaction
      transaction = oracleConnection.BeginTransaction();
      // Assign transaction object for a pending local transaction
      command.Transaction = transaction;

      try
      {
        updateGestaoCarga(gestaoCargaModel, command);
        transaction.Commit();
      }
      catch (Exception ex)
      {
        transaction.Rollback();

        throw ex;
      }
    }
  }
}

File: Backend/Administracao.ConciliacaoAPI/Code/GestaoCarga/GestaoCargaDAL.vb
Imports Oracle.ManagedDataAccess.Client
Imports Administracao.ConciliacaoAPI.Model
Imports Administracao.ConciliacaoAPI.Util

Namespace Code.GestaoCarga
  Public Class GestaoCargaDAL

    Public oracleConnection As OracleConnection
    Public Sub New()
      Try
        oracleConnection = New OracleConnection(Utils.ORACLE_CONNECCAO)
        oracleConnection.Open()
      Catch ex As Exception
        Throw New Exception("Erro ao conectar ao banco de dados.")
      End Try
    End Sub


    Public Function selectGestaoCargaPorId(id As Integer) As GestaoCargaModel

      Dim sql As New GestaoCargaSQL()


      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      command.CommandText = sql.selectGestaoCargaPorId()

      command.Parameters.Add("codigo", id)
      command.CommandType = System.Data.CommandType.Text
      Dim reader As OracleDataReader = command.ExecuteReader()
      Dim gestaoCarga As GestaoCargaModel = Nothing

      Try
        If reader.Read() Then
                    gestaoCarga = New GestaoCargaModel() With {
              .codGestaoCarga = reader.GetInt32(0),
              .dataMovimentacao = reader.GetDateTime(1),
              .data = reader.GetDateTime(2),
              .tipo = reader.GetString(3),
              .versao = reader.GetInt32(4),
              .versaoOficial = reader.GetString(5),
              .valor = reader.GetFloat(6),
               .versaoOficialLabel = If(reader.GetString(5) = "1", "Sim", "Não")
          }
                End If


        Return gestaoCarga
      Catch ex As Exception
        Throw ex
      Finally

        If oracleConnection IsNot Nothing Then
          oracleConnection.Close()
        End If
      End Try
    End Function
    Public Sub deleteGestaoCarga(id As Integer)

      Dim command As New OracleCommand()
      Dim transaction As OracleTransaction = oracleConnection.BeginTransaction()
      ' Assign transaction object for a pending local transaction
      command.Connection = oracleConnection
      command.Transaction = transaction
      Try
        Dim sql As New GestaoCargaSQL()
        command.CommandText = sql.deleteGestaoCargaCtb()
        command.Parameters.Add("codigo", id)

        ' LogUtil.Execute(New Dictionary(Of String, Object)(), "MOVCTBCLOSISCOB", command)

        command.ExecuteNonQuery()
        command.CommandText = sql.deleteGestaoCargaTit()
        command.ExecuteNonQuery()
        command.CommandText = sql.deleteGestaoCargaAgn()
        command.ExecuteNonQuery()
        command.CommandText = sql.deleteGestaoCarga()
        command.ExecuteNonQuery()
        transaction.Commit()
      Catch ex As Exception
        transaction.Rollback()

        Throw ex
      End Try
    End Sub

    Public Function selectGestaoCarga(dataRef As DateTime, tipo As [String]) As List(Of GestaoCargaModel)
      Dim sql As New GestaoCargaSQL()


      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      command.CommandText = sql.selectGestaoCarga()

      command.Parameters.Add("dataref", dataRef.ToString("yyyy-MM-dd"))
      command.Parameters.Add("tipo", tipo)
      command.CommandType = System.Data.CommandType.Text
      Dim reader As OracleDataReader = command.ExecuteReader()
      Dim gestaoCarga As New List(Of GestaoCargaModel)()

      Try

        While reader.Read()
                    gestaoCarga.Add(New GestaoCargaModel() With {
                        .codGestaoCarga = reader.GetInt32(0),
                        .dataMovimentacao = reader.GetDateTime(1),
                        .data = reader.GetDateTime(2),
                        .tipo = reader.GetString(3),
                        .versao = reader.GetInt32(4),
                        .versaoOficial = reader.GetString(5),
                        .versaoOficialLabel = If(reader.GetString(5) = "1", "Sim", "Não"),
                        .codUnidade = reader.GetInt32(6),
                        .CodConta = reader.GetInt32(7),
                        .valor = reader.GetFloat(8),
                        .quantidade = reader.GetInt32(9)
                    })
                End While


        Return gestaoCarga
      Catch ex As Exception
        Throw ex
      Finally

        If oracleConnection IsNot Nothing Then
          oracleConnection.Close()
        End If
      End Try
    End Function

    Private Sub updateGestaoCarga(gestaoCargaModel As GestaoCargaModel, command As OracleCommand)

      Dim sql As New GestaoCargaSQL()

      command.CommandText = sql.updateGestaoCargaVersaoOficialAtivar()
      command.Parameters.Clear()
      command.Parameters.Add("codigoUpdate", gestaoCargaModel.codGestaoCarga)

      command.ExecuteNonQuery()
      command.CommandText = sql.updateGestaoCargaVersaoOficialInativar()
      command.Parameters.Add("datarefUpdate", gestaoCargaModel.dataMovimentacao.ToString("yyyy-MM-dd"))
      command.Parameters.Add("tipoUpdate", gestaoCargaModel.tipo)
      command.ExecuteNonQuery()

      'Dim dic = New Dictionary(Of String, Object)()
      'dic.Add("IDTVRSRGTATV", 0)
      'LogUtil.Execute(dic, "RLCRGTMOVCLO", command)

      'command.ExecuteNonQuery()

    End Sub

    Public Sub insertGestaoCarga(dataRef As DateTime, tipo As String, codUnidade As Integer, codConta As Integer)

      If "TIT".Equals(tipo) OrElse "TODOS".Equals(tipo) Then
        inserirLoteTitulo(dataRef, codUnidade, codConta)
      End If
      If "CTB".Equals(tipo) OrElse "TODOS".Equals(tipo) Then
        inserirLoteContabil(dataRef, codUnidade, codConta)
      End If
      If "AGI".Equals(tipo) OrElse "TODOS".Equals(tipo) Then
        inserirLoteAging(dataRef, codUnidade, codConta)
      End If

    End Sub
    Public Function proximoCodigoGestaoCarga(command As OracleCommand) As Integer
      Dim sql As New GestaoCargaSQL()
      command.CommandText = sql.proximoCodigoGestaoCarga()
      command.CommandType = System.Data.CommandType.Text
      Dim reader As OracleDataReader = command.ExecuteReader()
      Dim codGestaoCarga As Integer = 0
      If reader.Read() Then
        codGestaoCarga = reader.GetInt32(0)
      End If
      Return codGestaoCarga
    End Function
    Private Sub inserirLoteAging(dataRef As DateTime, codUnidade As Integer, codConta As Integer)
      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      Dim transaction As OracleTransaction

      ' Start a local transaction
      transaction = oracleConnection.BeginTransaction()
      ' Assign transaction object for a pending local transaction
      command.Transaction = transaction

      Try

        Dim sql As New GestaoCargaSQL()

        Dim gestaoCargaModel As GestaoCargaModel = gerarGestaoCargaModel(dataRef, "AGI", codUnidade, codConta, command)


        command.CommandText = sql.insertLoteAging()
        command.CommandType = System.Data.CommandType.Text
        command.Parameters.Clear()
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga)
        command.Parameters.Add("datamov", gestaoCargaModel.dataMovimentacao)
        command.ExecuteNonQuery()
        transaction.Commit()
      Catch ex As Exception
        transaction.Rollback()

        Throw ex
      End Try
    End Sub
    Public Function gerarGestaoCargaModel(dataref As DateTime, tipo As String, codUnidade As Integer, codConta As Integer, command As OracleCommand) As GestaoCargaModel

      Dim gestaoCargaModel As New GestaoCargaModel()
      gestaoCargaModel.codGestaoCarga = proximoCodigoGestaoCarga(command)
      gestaoCargaModel.tipo = tipo
      gestaoCargaModel.data = DateTime.Now
      gestaoCargaModel.dataMovimentacao = dataref
      gestaoCargaModel.codUnidade = codUnidade
      gestaoCargaModel.CodConta = codConta
      gestaoCargaModel.versao = selectMaxVersao(dataref, gestaoCargaModel.tipo, command)
      Dim sql As New GestaoCargaSQL()
      command.CommandText = sql.insertGestaoCarga()
      command.Parameters.Clear()
      command.Parameters.Add("NUMSEQRGTCLO", gestaoCargaModel.codGestaoCarga)
      command.Parameters.Add("DATREFRGTCLO", gestaoCargaModel.dataMovimentacao)
      command.Parameters.Add("TIPRGTCLO", gestaoCargaModel.tipo)
      command.Parameters.Add("NUMVRSRGTCLO", gestaoCargaModel.versao)
      command.Parameters.Add("CODUNDNGC", gestaoCargaModel.codUnidade)
      command.Parameters.Add("CODCNTCTB", gestaoCargaModel.CodConta)
      command.ExecuteNonQuery()
      updateGestaoCarga(gestaoCargaModel, command)
      Return gestaoCargaModel
    End Function

    Private Function selectMaxVersao(data As DateTime, tipo As String, command As OracleCommand) As Integer

      Dim sql As New GestaoCargaSQL()
      command.CommandText = sql.selectMaxVersion()
      command.CommandType = System.Data.CommandType.Text
      command.Parameters.Add("dataRefVer", data.ToString("yyyy-MM-dd"))
      command.Parameters.Add("tipoVer", tipo)
      Dim reader As OracleDataReader = command.ExecuteReader()
      Dim versao As Integer = 0
      If reader.Read() Then
        versao = reader.GetInt32(0)
      End If
      Return versao
    End Function

    Private Sub inserirLoteContabil(dataRef As DateTime, codUnidade As Integer, codConta As Integer)

      Dim transaction As OracleTransaction

      ' Start a local transaction
      transaction = oracleConnection.BeginTransaction()
      ' Assign transaction object for a pending local transaction
      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      command.Transaction = transaction

      Try
        Dim sql As New GestaoCargaSQL()
        Dim gestaoCargaModel As GestaoCargaModel = gerarGestaoCargaModel(dataRef, "CTB", codUnidade, codConta, command)

        command.CommandText = sql.insertLoteContabil()
        command.CommandType = System.Data.CommandType.Text
        command.Parameters.Clear()
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga)
        command.Parameters.Add("datamov", gestaoCargaModel.dataMovimentacao)
        command.ExecuteNonQuery()
        transaction.Commit()
      Catch ex As Exception
        transaction.Rollback()

        Throw ex
      End Try
    End Sub

    Private Sub inserirLoteTitulo(dataRef As DateTime, codUnidade As Integer, codConta As Integer)


      Dim transaction As OracleTransaction

      ' Start a local transaction
      transaction = oracleConnection.BeginTransaction()
      ' Assign transaction object for a pending local transaction
      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      command.Transaction = transaction

      Try
        Dim sql As New GestaoCargaSQL()

        Dim gestaoCargaModel As GestaoCargaModel = gerarGestaoCargaModel(dataRef, "TIT", codUnidade, codConta, command)

        command.CommandText = sql.insertLoteTitulo()
        command.CommandType = System.Data.CommandType.Text
        command.Parameters.Clear()
        command.Parameters.Add("Codigo", gestaoCargaModel.codGestaoCarga)
        command.Parameters.Add("datamov", gestaoCargaModel.dataMovimentacao)
        command.ExecuteNonQuery()

        ' LogUtil.Execute(New Dictionary(Of String, Object)(), "MOVTITCLOSISCOB", command, "where NUMSEQCRGTAB = :Codigo")

        transaction.Commit()
      Catch ex As Exception
        transaction.Rollback()

        Throw ex
      End Try
    End Sub
    Public Sub atualizarVersao(gestaoCargaModel As GestaoCargaModel)
      Dim command As New OracleCommand()
      command.Connection = oracleConnection
      Dim transaction As OracleTransaction

      ' Start a local transaction
      transaction = oracleConnection.BeginTransaction()
      ' Assign transaction object for a pending local transaction
      command.Transaction = transaction

      Try
        updateGestaoCarga(gestaoCargaModel, command)
        transaction.Commit()
      Catch ex As Exception
        transaction.Rollback()

        Throw ex
      End Try
    End Sub
  End Class
End Namespace
